src/java.base/unix/classes/sun/nio/fs/UnixPath.java

```diff
@@ -857,13 +860,76 @@ public Path toRealPath(LinkOption... options) throws IOException {
             result = result.resolve(element);
         }
 
-        // check file exists (without following links)
+        // check whether file exists (without following links)
         try {
             UnixFileAttributes.get(result, false);
         } catch (UnixException x) {
             x.rethrowAsIOException(result);
         }
-        return result;
+
+        // Return if the file system is not both case insensitive and retentive
+        if (!fs.isCaseInsensitiveAndPreserving())
+            return result;
+
+        UnixPath path = fs.rootDirectory();
+
+        SecurityManager sm = System.getSecurityManager();
+
+        // Traverse the result obtained above from the root downward, leaving
+        // any '..' elements intact, and replacing other elements with the
+        // entry in the same directory which has an equal key
+        for (int i = 0; i < result.getNameCount(); i++ ) {
+            UnixPath element = result.getName(i);
+
+            // If the element is "..", append it directly and continue
+            if (element.toString().equals("..")) {
+                path = path.resolve(element);
+                continue;
+            }
+
+            // Derive full path to element and check readability
+            UnixPath elementPath = path.resolve(element);
+
+            // Derive element key
+            UnixFileAttributes attrs = null;
+            try {
+                attrs = UnixFileAttributes.get(elementPath, false);
+            } catch (UnixException x) {
+                x.rethrowAsIOException(result);
+            }
+            final UnixFileKey elementKey = attrs.fileKey();
+
+            // Obtain the stream of entries in the directory corresponding
+            // to the path constructed thus far, and extract the entry whose
+            // key is equal to the key of the current element
+            DirectoryStream.Filter<Path> filter = (p) -> { return true; };
+            try (DirectoryStream<Path> entries =
+                getFileSystem().provider().newDirectoryStream(path, filter)) {
```

Comment:
I think the algorithm is correct, just need to fix the strange intend at 907.

Reply:
> I think the algorithm is correct, just need to fix the strange intend at 907.

I assume you meant _indent(ation)_. If so, then I wonder what you'd consider as acceptable formatting here. Could `var` help not solve this problem by making 906-907 a one-liner instead? Or this would be too unusual for this class?

Reply:
Indentation modified in c193b3b48f56e2c581efaf212f2770c9add36a0a. There is no use of `var` in this class so far, thus although it is a good idea, `var` is not used here, either.

Reply:
> If so, then I wonder what you'd consider as acceptable formatting here. Could `var` help not solve this problem by making 906-907 a one-liner instead? Or this would be too unusual for this class?

The main thing when reading this code is to satisfied that it implements the scan correctly and to see how it behaves when it can't open a directory or read the attributes of one of the entries in the directory. The issue with the version that were discussing is that it wasn't immediately obvious which code was in the try resources specification vs. try block. He's changed the indent so it is more obvious now, personally I would have put it one line and renamed it to "provider" so that it's the same name used elsewhere in this code.